regolith/file_system.go

```diff
@@ -872,3 +873,110 @@ func MoveOrCopy(
 	}
 	return nil
 }
+
+// SyncDirectories copies the source to destination while checking size and modification time.
+func SyncDirectories(
+	source string, destination string, makeReadOnly bool, copyParentAcl bool,
+) error {
+	// Make destination parent if not exists
+	destinationParent := filepath.Dir(destination)
+	if _, err := os.Stat(destinationParent); os.IsNotExist(err) {
+		err = os.MkdirAll(destinationParent, 0755)
+		if err != nil {
+			return burrito.WrapErrorf(
+				err, osMkdirError, destinationParent)
+		}
+	}
+	err := filepath.Walk(source, func(srcPath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		relPath, err := filepath.Rel(source, srcPath)
+		if err != nil {
+			return burrito.WrapErrorf(err, filepathRelError, source, srcPath)
+		}
+		destPath := filepath.Join(destination, relPath)
+
+		destInfo, err := os.Stat(destPath)
+		if err != nil && !os.IsNotExist(err) {
+			return burrito.WrapErrorf(err, osStatErrorAny, destPath)
+		}
+		if (err != nil && os.IsNotExist(err)) || info.ModTime() != destInfo.ModTime() || info.Size() != destInfo.Size() {
+			if info.IsDir() {
+				return os.MkdirAll(destPath, info.Mode())
+			}
+			Logger.Debugf("SYNC: Copying file %s to %s", srcPath, destPath)
+			return copyFile(srcPath, destPath, info)
+		} else {
+			Logger.Debugf("SYNC: Skipping file %s", srcPath)
+		}
+		return nil
+	})
+	if err != nil {
+		return burrito.WrapErrorf(err, osCopyError, source, destination)
+	}
+
+	// Remove files/folders in destination that are not in source
+	err = filepath.Walk(destination, func(destPath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		relPath, err := filepath.Rel(destination, destPath)
+		if err != nil {
+			return burrito.WrapErrorf(err, filepathRelError, destination, destPath)
+		}
+		srcPath := filepath.Join(source, relPath)
+		if _, err := os.Stat(srcPath); os.IsNotExist(err) {
+			Logger.Debugf("SYNC: Removing file %s", destPath)
+			return os.RemoveAll(destPath)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return burrito.WrapErrorf(err, osRemoveError, source, destination)
+	}
+
+	//TODO: copy ACL. To be honest, I have no clue why it was there in the first place
+
+	// Make files read only if this option is selected
+	if makeReadOnly {
```

Comment:
Does changing the read only property update the last modification time of the file? If yes, would that cause issues with this function?

Reply:
Tbh I did not test whether making read only changes the modification time. I'll check it.

Reply:
I don't think it does but there was an issue that caused the syncing function to crash when it encountered read-only files. I fixed it.